src/Shared/PackageManagers/MavenProjectManager.cs

```diff
@@ -501,28 +393,112 @@ private static MavenArtifactType GetMavenArtifactType(string fileName)
                 // Split the version content into its individual parts.
                 // [0] - The version
                 // [1] - The date it was published
-                // [2] - The time it waas published
-                // [3] - The download count
+                // [2] - The time it was published
+                // [3] - The file size in bytes
                 string[] versionParts = versionContent.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                 if (DateTime.TryParse($"{versionParts[1]} {versionParts[2]}", out DateTime publishDateTime))
                 {
                     return publishDateTime;
                 }
             }
-            else if (upstream == MavenSupportedUpstream.GoogleMavenRepository)
+
+            // If the directory index approach does not work, try to get the "Last-Modified" header from the .pom file.
+            baseUrl = $"{feedUrl.EnsureTrailingSlash()}{packageNamespace}/{packageName}/{packageVersion}/{packageName}-{packageVersion}.pom";
+            HttpResponseMessage response = await httpClient.GetAsync(baseUrl);
+
+            if (response.Content.Headers.TryGetValues("Last-Modified", out var values))
+            {
+                var lastModified = DateTime.Parse(values.First());
+                return lastModified;
+            }
+
+            return null;
+        }
+
+        private async IAsyncEnumerable<string> GetArtifactDownloadUris_DirectoryIndexStrategyAsync(string baseUrl, bool useCache = true)
+        {
+            HttpClient httpClient = CreateHttpClient();
+            var fileNameList = new List<string>();
+
+            try
+            {
+                string? html = await GetHttpStringCache(httpClient, baseUrl, useCache);
+                if (string.IsNullOrEmpty(html))
+                {
+                    throw new InvalidOperationException();
+                }
+
+                HtmlParser parser = new();
+                AngleSharp.Html.Dom.IHtmlDocument document = await parser.ParseDocumentAsync(html);
+
+                foreach (string fileName in document.QuerySelectorAll("a").Select(link => link.GetAttribute("href").ToString()))
+                {
+                    if (fileName == "../") continue;
+                    fileNameList.Add(baseUrl + fileName);
+                }
+            }
+            catch {}
+
+            foreach (var file in fileNameList)
             {
-                var baseUrl = $"{upstream.GetDownloadRepositoryUrl().EnsureTrailingSlash()}{packageNamespace}/{packageName}/{packageVersion}/{packageName}-{packageVersion}.pom";
-                HttpResponseMessage response = await httpClient.GetAsync(baseUrl);
+                yield return file;
+            }
```

Comment:
Nit: As I mentioned before, `IAsyncEnumerable` appears to be overkill for methods like these. As I understand, the primary value add of `IAsyncEnumerable` is that one could add statements between calls to `yield` that need to `await` on something to get more results; like `await`ing on a network call to get another page of results before yielding each item in each page.

When we already have everything in one shot, a simple `IEnumerable` might suffice. We might even be able to return the LINQ expression result directly, something like this:

```c#
return document.QuerySelectorAll("a")
  .Select(link => link.GetAttribute("href").ToString())
  .Where(fileName != "../")
  .Select(fileName => baseUrl + fileName);
```

Reply:
Same note applies to the other strategies as well.

Reply:
Okay, good to know. I changed the return type to `Task<List<string>>?` to simplify this and differentiate between unsuccessful vs. successful with no results.